plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java

```diff
@@ -0,0 +1,245 @@
+            ConnectorSession session,
+            PhoenixOutputTableHandle phoenixOutputTableHandle,
+            PhoenixClient phoenixClient,
+            ConnectorPageSinkId pageSinkId,
+            RemoteQueryModifier remoteQueryModifier)
+    {
+        ImmutableList.Builder<String> columnNamesBuilder = ImmutableList.builder();
+        ImmutableList.Builder<Type> columnTypesBuilder = ImmutableList.builder();
+        columnNamesBuilder.addAll(phoenixOutputTableHandle.getColumnNames());
+        columnTypesBuilder.addAll(phoenixOutputTableHandle.getColumnTypes());
+        if (hasRowKey) {
+            columnNamesBuilder.add(ROWKEY);
+            columnTypesBuilder.add(ROWKEY_COLUMN_HANDLE.getColumnType());
+        }
+
+        PhoenixOutputTableHandle updateOutputTableHandle = new PhoenixOutputTableHandle(
+                schemaName,
+                tableName,
+                columnNamesBuilder.build(),
+                columnTypesBuilder.build(),
+                Optional.empty(),
+                Optional.empty());
+        return new JdbcPageSink(session, updateOutputTableHandle, phoenixClient, pageSinkId, remoteQueryModifier);
+    }
+
+    private ConnectorPageSink createDeleteSink(
+            ConnectorSession session,
+            List<Type> mergeRowIdFieldTypes,
+            PhoenixClient phoenixClient,
+            ConnectorPageSinkId pageSinkId,
+            RemoteQueryModifier remoteQueryModifier)
+    {
+        checkArgument(mergeRowIdFieldNames.size() == mergeRowIdFieldTypes.size(), "Wrong merge row column, columns and types size not match");
+        JdbcOutputTableHandle deleteOutputTableHandle = new PhoenixOutputTableHandle(
+                schemaName,
+                tableName,
+                mergeRowIdFieldNames,
+                mergeRowIdFieldTypes,
+                Optional.empty(),
+                Optional.empty());
+
+        return new DeleteSink(session, deleteOutputTableHandle, phoenixClient, pageSinkId, remoteQueryModifier);
+    }
+
+    private class DeleteSink
+            extends JdbcPageSink
+    {
+        public DeleteSink(ConnectorSession session, JdbcOutputTableHandle handle, JdbcClient jdbcClient, ConnectorPageSinkId pageSinkId, RemoteQueryModifier remoteQueryModifier)
+        {
+            super(session, handle, jdbcClient, pageSinkId, remoteQueryModifier);
+        }
+
+        @Override
+        protected String getSinkSql(JdbcClient jdbcClient, JdbcOutputTableHandle outputTableHandle, List<WriteFunction> columnWriters)
+        {
+            List<String> conjuncts = mergeRowIdFieldNames.stream()
+                    .map(name -> name + " = ? ")
+                    .collect(toImmutableList());
+            checkArgument(!conjuncts.isEmpty(), "Merge row id fields should not empty");
+            String whereCondition = Joiner.on(" AND ").join(conjuncts);
+
+            return format("DELETE FROM %s.%s WHERE %s", schemaName, tableName, whereCondition);
```

Comment:
does this work if the table is non-lowercase?

Reply:
Yes. Changed the table name to upper case in the `testMergeWithSpecifiedRowkeys` test.